Allow custom ports for DBs in serverpod create

TylerSustare opened the pull request:
## Closes https://github.com/serverpod/serverpod/issues/523

[Adds the ability to customize DB ports](https://github.com/serverpod/serverpod/pull/532/commits/46d9599d0bc71a0fe21b82630face83a061dacb0) 

This will allow users with busy ports on 8090 and 8091 to use serverpod
without the `-f` option when creating a new project.

This is the simplest way to allow a user to eject from these ports
while keeping the ability to run the "Templates" as individual
projects.
## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update cointains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [ ] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

## Breaking changes

_If you have done any breaking changes, make sure to outline them here, so that they can be included in the notes for the next release._

TylerSustare commented:
Following https://github.com/serverpod/serverpod/pull/526

vlidholt commented:
A consideration which makes this feature a bit more tricky to implement is that the template files need to be runnable projects by themselves. Otherwise, it will be much harder to test the templates. (As running a test would require running serverpod create first.)

TylerSustare commented:
I'm moving the PG error into a comment instead of the description.

____
Currently, there is a bug with creating the DB user. I'm actively working on this, which is why it's still in a draft state. 

currently the command
```
➜ dart serverpod/tools/serverpod_cli/bin/serverpod_cli.dart create --postgresPort=7777 serverpod_test
```
Gets this error
```
Postgres is ready
psql: error: connection to server at "postgres" (172.31.0.3), port 5432 failed: FATAL:  password authentication failed for user "postgres"
Stopping docker
```

Full output
<details>
➜ dart serverpod/tools/serverpod_cli/bin/serverpod_cli.dart create --postgresPort=7777 serverpod_test
Development mode. Using templates from: /Users/tyler/go/src/github.com/tylersustare/serverpod/templates/serverpod_templates
SERVERPOD_HOME is set to /Users/tyler/go/src/github.com/tylersustare/serverpod
Creating project serverpod_test...
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/generated/tables-serverpod.pgsql
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/generated/protocol.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/generated/tables.pgsql
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/docker-compose.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/bin/main.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/CHANGELOG.md
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/setup-tables.cmd
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/config/generator.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/config/staging.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/config/production.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/config/development.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/config/passwords.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/web/static/css/style.css
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/web/static/images/background.svg
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/web/static/images/serverpod-logo.svg
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/web/templates/default.html
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/README.md
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/pubspec.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/setup-tables
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/server.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/generated/endpoints.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/generated/protocol.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/generated/example.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/endpoints/example_endpoint.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/future_calls/example_future_call.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/web/routes/root.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/web/widgets/default_page_widget.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/lib/src/protocol/example.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/analysis_options.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/cloudfront-web.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/cloudfront-web-staging.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/staging.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/config.auto.tfvars
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/main.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/storage.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/init-script.sh
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/redis.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/balancers.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/database.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/variables.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/code-deploy.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/instances.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/balancers-staging.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/gitignore
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/terraform/vpc.tf
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/scripts/start_server
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/scripts/appspec.yml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/scripts/run_serverpod
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/aws/scripts/install_dependencies
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server/gitignore
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/CHANGELOG.md
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/README.md
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/pubspec.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/lib/projectname_client.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/lib/src/protocol/client.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/lib/src/protocol/protocol.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/lib/src/protocol/example.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/analysis_options.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client/gitignore
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter/test/widget_test.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter/README.md
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter/pubspec.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter/lib/main.dart
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter/analysis_options.yaml
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter/gitignore
  /Users/tyler/go/src/github.com/tylersustare/serverpod_test/.github/workflows/deployment-aws.yml

Running `dart pub get` in /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server
Resolving dependencies...
+ args 2.3.1
+ async 2.10.0
+ buffer 1.1.1
+ collection 1.17.0
+ crypto 3.0.2
+ executor 2.2.2
+ http 0.13.5
+ http_parser 4.0.2
+ lints 2.0.1
+ meta 1.8.0
+ mustache_template 2.0.0
+ path 1.8.3
+ pedantic 1.11.1
+ postgres 2.5.2
+ postgres_pool 2.1.5
+ redis 3.1.0
+ retry 3.1.0
+ sasl_scram 0.1.1
+ saslprep 1.0.2
+ serverpod 0.9.21
+ serverpod_client 0.9.21
+ serverpod_serialization 0.9.21
+ serverpod_service_client 0.9.21
+ serverpod_shared 0.9.21
+ source_span 1.9.1
+ stack_trace 1.11.0
+ stream_channel 2.1.1
+ string_scanner 1.2.0
+ synchronized 3.0.0+3
+ system_resources 1.6.0
+ term_glyph 1.2.1
+ typed_data 1.3.1
+ unorm_dart 0.2.0
+ vm_service 9.4.0
+ web_socket_channel 2.2.0
+ yaml 3.1.1
Changed 36 dependencies!

Running `dart pub get` in /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_client
Resolving dependencies...
+ async 2.10.0
+ collection 1.17.0
+ crypto 3.0.2
+ http 0.13.5
+ http_parser 4.0.2
+ meta 1.8.0
+ path 1.8.3
+ serverpod_client 0.9.21
+ serverpod_serialization 0.9.21
+ source_span 1.9.1
+ stream_channel 2.1.1
+ string_scanner 1.2.0
+ term_glyph 1.2.1
+ typed_data 1.3.1
+ web_socket_channel 2.2.0
+ yaml 3.1.1
Changed 16 dependencies!

Running `flutter create .` in /Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_flutter
Recreating project ....
  windows/runner/flutter_window.cpp (created)
  windows/runner/utils.h (created)
  windows/runner/utils.cpp (created)
  windows/runner/runner.exe.manifest (created)
  windows/runner/CMakeLists.txt (created)
  windows/runner/win32_window.h (created)
  windows/runner/Runner.rc (created)
  windows/runner/win32_window.cpp (created)
  windows/runner/resources/app_icon.ico (created)
  windows/runner/main.cpp (created)
  windows/runner/resource.h (created)
  windows/runner/flutter_window.h (created)
  windows/flutter/CMakeLists.txt (created)
  windows/.gitignore (created)
  windows/CMakeLists.txt (created)
  ios/Runner.xcworkspace/contents.xcworkspacedata (created)
  ios/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (created)
  ios/Runner.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings (created)
  ios/Runner/Info.plist (created)
  ios/Runner/Assets.xcassets/LaunchImage.imageset/LaunchImage@2x.png (created)
  ios/Runner/Assets.xcassets/LaunchImage.imageset/LaunchImage@3x.png (created)
  ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md (created)
  ios/Runner/Assets.xcassets/LaunchImage.imageset/Contents.json (created)
  ios/Runner/Assets.xcassets/LaunchImage.imageset/LaunchImage.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-76x76@2x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-29x29@1x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-40x40@1x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-20x20@1x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-1024x1024@1x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-83.5x83.5@2x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-20x20@3x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Contents.json (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-20x20@2x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-29x29@3x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-40x40@2x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-60x60@3x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-60x60@2x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-76x76@1x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-40x40@3x.png (created)
  ios/Runner/Assets.xcassets/AppIcon.appiconset/Icon-App-29x29@2x.png (created)
  ios/Runner/Base.lproj/LaunchScreen.storyboard (created)
  ios/Runner/Base.lproj/Main.storyboard (created)
  ios/Runner.xcodeproj/project.xcworkspace/contents.xcworkspacedata (created)
  ios/Runner.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (created)
  ios/Runner.xcodeproj/project.xcworkspace/xcshareddata/WorkspaceSettings.xcsettings (created)
  ios/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme (created)
  ios/Flutter/Debug.xcconfig (created)
  ios/Flutter/Release.xcconfig (created)
  ios/Flutter/AppFrameworkInfo.plist (created)
  ios/.gitignore (created)
  serverpod_test_flutter.iml (created)
  web/favicon.png (created)
  web/index.html (created)
  web/manifest.json (created)
  web/icons/Icon-maskable-512.png (created)
  web/icons/Icon-192.png (created)
  web/icons/Icon-maskable-192.png (created)
  web/icons/Icon-512.png (created)
  macos/Runner.xcworkspace/contents.xcworkspacedata (created)
  macos/Runner.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_32.png (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_16.png (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_256.png (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_64.png (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_512.png (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/Contents.json (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_128.png (created)
  macos/Runner/Assets.xcassets/AppIcon.appiconset/app_icon_1024.png (created)
  macos/Runner/DebugProfile.entitlements (created)
  macos/Runner/Base.lproj/MainMenu.xib (created)
  macos/Runner/MainFlutterWindow.swift (created)
  macos/Runner/Configs/Debug.xcconfig (created)
  macos/Runner/Configs/Release.xcconfig (created)
  macos/Runner/Configs/Warnings.xcconfig (created)
  macos/Runner/Configs/AppInfo.xcconfig (created)
  macos/Runner/AppDelegate.swift (created)
  macos/Runner/Info.plist (created)
  macos/Runner/Release.entitlements (created)
  macos/Runner.xcodeproj/project.xcworkspace/xcshareddata/IDEWorkspaceChecks.plist (created)
  macos/Runner.xcodeproj/project.pbxproj (created)
  macos/Runner.xcodeproj/xcshareddata/xcschemes/Runner.xcscheme (created)
  macos/Flutter/Flutter-Debug.xcconfig (created)
  macos/Flutter/Flutter-Release.xcconfig (created)
  macos/.gitignore (created)
  android/app/src/profile/AndroidManifest.xml (created)
  android/app/src/main/res/mipmap-mdpi/ic_launcher.png (created)
  android/app/src/main/res/mipmap-hdpi/ic_launcher.png (created)
  android/app/src/main/res/drawable/launch_background.xml (created)
  android/app/src/main/res/mipmap-xxxhdpi/ic_launcher.png (created)
  android/app/src/main/res/mipmap-xxhdpi/ic_launcher.png (created)
  android/app/src/main/res/values-night/styles.xml (created)
  android/app/src/main/res/values/styles.xml (created)
  android/app/src/main/res/drawable-v21/launch_background.xml (created)
  android/app/src/main/res/mipmap-xhdpi/ic_launcher.png (created)
  android/app/src/main/AndroidManifest.xml (created)
  android/app/src/debug/AndroidManifest.xml (created)
  android/gradle/wrapper/gradle-wrapper.properties (created)
  android/gradle.properties (created)
  android/.gitignore (created)
  android/settings.gradle (created)
  android/app/build.gradle (created)
  android/app/src/main/kotlin/com/example/serverpod_test_flutter/MainActivity.kt (created)
  android/build.gradle (created)
  android/serverpod_test_flutter_android.iml (created)
  ios/Runner/Runner-Bridging-Header.h (created)
  ios/Runner/AppDelegate.swift (created)
  ios/Runner.xcodeproj/project.pbxproj (created)
  .idea/runConfigurations/main_dart.xml (created)
  .idea/libraries/Dart_SDK.xml (created)
  .idea/libraries/KotlinJavaRuntime.xml (created)
  .idea/modules.xml (created)
  .idea/workspace.xml (created)
  linux/main.cc (created)
  linux/my_application.h (created)
  linux/my_application.cc (created)
  linux/flutter/CMakeLists.txt (created)
  linux/.gitignore (created)
  linux/CMakeLists.txt (created)
Running "flutter pub get" in serverpod_test_flutter...           1,401ms
Wrote 121 files.

All done!
In order to run your application, type:

  $ cd .
  $ flutter run

Your application code is in ./lib/main.dart.


Setting up Docker and default database tables in
/Users/tyler/go/src/github.com/tylersustare/serverpod_test/serverpod_test_server
If you run serverpod create for the first time, this can take a few minutes as Docker is downloading the
images for Postgres. If you get stuck at this step, make sure that you have the latest version of Docker
Desktop and that it is currently running.

Starting docker
Network serverpod_test_server_default  Creating
Network serverpod_test_server_default  Created
Container serverpod_test_server-redis-1  Creating
Container serverpod_test_server-postgres-1  Creating
Container serverpod_test_server-postgres-1  Created
Container serverpod_test_server-redis-1  Created
Container serverpod_test_server-redis-1  Starting
Container serverpod_test_server-postgres-1  Starting
Container serverpod_test_server-postgres-1  Started
Container serverpod_test_server-redis-1  Started
Postgres is ready
psql: error: connection to server at "postgres" (172.31.0.3), port 5432 failed: FATAL:  password authentication failed for user "postgres"
Stopping docker
Container serverpod_test_server-redis-1  Stopping
Container serverpod_test_server-postgres-1  Stopping
Container serverpod_test_server-postgres-1  Stopped
Container serverpod_test_server-redis-1  Stopped
Completed table setup exit code: 0
Cleaning up




SERVERPOD CREATED 🥳

All setup. You are ready to rock!

Start your Serverpod by running:

  $ cd serverpod_test/serverpod_test_server
  $ docker-compose up --build --detach
  $ dart bin/main.dart
</details>

Running the server is producing the expected error after not being able to create a DB user.
```
SERVERPOD version: 0.9.21 mode: development time: 2022-12-30 00:27:48.306228Z
2022-12-30 00:27:48.876964Z Failed to connect to database.
PostgreSQLSeverity.fatal 28P01: password authentication failed for user "postgres"
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 331:18                 PgPool.run.<fn>
package:retry/retry.dart 131:16                                 RetryOptions.retry
package:postgres_pool/postgres_pool.dart 329:14                 PgPool.run
package:serverpod/src/database/database_connection.dart 171:20  DatabaseConnection.find
package:serverpod/src/database/database_connection.dart 199:18  DatabaseConnection.findSingleRow
package:serverpod/src/database/database.dart 86:12              Database.findSingleRow
package:serverpod/src/server/serverpod.dart 310:13              Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 299:5               Serverpod.start
package:serverpod_test_server/server.dart 33:3                  run

2022-12-30 00:27:50.310686Z Internal server error. Failed to load method lookup.
PostgreSQLSeverity.fatal 28P01: password authentication failed for user "postgres"
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 331:18                 PgPool.run.<fn>
package:retry/retry.dart 131:16                                 RetryOptions.retry
package:postgres_pool/postgres_pool.dart 329:14                 PgPool.run
package:serverpod/src/database/database_connection.dart 171:20  DatabaseConnection.find
package:serverpod/src/database/database_connection.dart 199:18  DatabaseConnection.findSingleRow
package:serverpod/src/database/database.dart 86:12              Database.findSingleRow
package:serverpod/src/server/method_lookup.dart 47:26           MethodLookup._attemptLoad
package:serverpod/src/server/method_lookup.dart 30:7            MethodLookup.load
package:serverpod/src/server/serverpod.dart 325:9               Serverpod.start.<fn>
package:serverpod/src/server/serverpod.dart 299:5               Serverpod.start
package:serverpod_test_server/server.dart 33:3                  run

Insights listening on port 8081
Server default listening on port 8080
Webserver listening on port 8082
2022-12-30 00:27:50.711029Z Internal server error. Failed to connect to database in future call manager.
PostgreSQLSeverity.fatal 28P01: password authentication failed for user "postgres"
===== asynchronous gap ===========================
package:postgres_pool/postgres_pool.dart 331:18                 PgPool.run.<fn>
package:retry/retry.dart 131:16                                 RetryOptions.retry
package:postgres_pool/postgres_pool.dart 329:14                 PgPool.run
package:serverpod/src/database/database_connection.dart 443:20  DatabaseConnection.deleteAndReturn
package:serverpod/src/database/database.dart 166:12             Database.deleteAndReturn
package:serverpod/src/server/future_call_manager.dart 100:18    FutureCallManager._checkQueue

Local stacktrace:
#0      FutureCallManager._checkQueue (package:serverpod/src/server/future_call_manager.dart:136:36)
<asynchronous suspension>
```

TylerSustare commented:
@vlidholt That seems fair. I'll work on that next.

The postgres error may not have anything to do with these changes though. As running this in a debug mode step-by-step avoided it. And I got this same error when creating a new project without a custom port in development mode.

I'm not able to test if this happens in the production mode since I can't create a serverpod project through the production CLI.

TylerSustare commented:
Yeah, this PG password issue isn't related to this PR. There are discussions about it happening, so I'm going to disregard it. https://github.com/serverpod/serverpod/discussions/121

I'll still make the changes so that
> template files need to be runnable projects by themselves.

TylerSustare commented on tools/serverpod_cli/bin/create/create.dart:
```diff
@@ -11,24 +11,29 @@ import '../util/print.dart';
+    'Serverpod API': 8080,
+    'Serverpod insights API': 8081,
+    'Serverpod Relic web server': 8082,
+    'PostgreSQL server': postgresPort,
+    'Redis server': redisPort,
```
These will always come in as the default ports or custom port because of the `defaultsTo` value in the CLI arg.

TylerSustare commented:
This is the most minimal way I can change the ports and is the only way I can use Serverpod because my machine is still using 8090 as the `launchd` port. 

I hope this isn't closed and immediately locked.

vlidholt commented on tools/serverpod_cli/bin/create/create.dart:
```diff
@@ -11,24 +11,29 @@ import '../util/print.dart';
-  'PostgreSQL server': 8090,
-  'Redis server': 8091,
-};
+const _defaultPostgresPort = '8090';
```
These should be defined as integers.

vlidholt commented on tools/serverpod_cli/bin/create/create.dart:
```diff
@@ -11,24 +11,29 @@ import '../util/print.dart';
   bool verbose,
   String template,
   bool force,
+  int postgresPort,
```
Make optional. Also, with so many parameters, named parameters will be a better choice.

vlidholt commented on tools/serverpod_cli/bin/serverpod_cli.dart:
```diff
@@ -115,6 +115,20 @@ Future<void> _main(List<String> args) async {
+  createParser.addOption(
+    'postgresPort',
+    abbr: 'p',
+    defaultsTo: '8090',
```
No need for `defaultsTo` if postgresPort is optional in `performCreate`. If using `defaultsTo`, the defined constant should be used.

TylerSustare replied:
sure

vlidholt commented on tools/serverpod_cli/bin/serverpod_cli.dart:
```diff
@@ -177,6 +191,8 @@ Future<void> _main(List<String> args) async {
       bool verbose = results.command!['verbose'];
       String template = results.command!['template'];
       bool force = results.command!['force'];
+      int pgPort = int.parse(results.command!['postgresPort']);
```
`postgresPort` for consistency. This should either result in a meaningful error message rather than throwing an exception on an invalid value. Or, a simpler solution could be to just use the default port, in which case you can use `int.tryParse`

vlidholt reviewed (changes requested):
This is a great addition, it will be approved with the suggested changes. Although, it must also make the setup script up-to-date so that database tables are correctly created. Or is that working already?

TylerSustare commented on tools/serverpod_cli/bin/create/create.dart:
```diff
@@ -11,24 +11,29 @@ import '../util/print.dart';
+    'Serverpod API': 8080,
+    'Serverpod insights API': 8081,
+    'Serverpod Relic web server': 8082,
+    'PostgreSQL server': postgresPort,
```
These will always be populated because of the default value in the CLI arg.

TylerSustare commented:
I'm going to close this PR. It hasn't been worth trying to get a "Hello World" running with this amount of effort. 
I haven't enjoyed working with this tool and I don't believe it's worth my time. Contributors can feel free to use this as a basis to get this feature running. But I'm out on Serverpod, which is too bad. I was really excited for this.
